cms/djangoapps/contentstore/views/helpers.py

```diff
@@ -271,6 +281,77 @@ def create_xblock(parent_locator, user, category, display_name, boilerplate=None
+        return self.context_key.make_usage_key(def_id.block_type, LocalId())
+
+    def create_definition(self, block_type, slug=None) -> DefinitionLocator:
+        """ Generate a new definition_id for an XBlock """
+        # Note: Split modulestore will detect this temporary ID and create a new definition ID when the XBlock is saved.
+        return DefinitionLocator(block_type, LocalId(block_type))
+
+
+def import_staged_content_from_user_clipboard(parent_key: UsageKey, user):
+    """
+    Import a block (and any children it has) from "staged" OLX.
+    Does not deal with permissions or REST stuff - do that before calling this.
+
+    Returns the newly created block on success or None if the clipboard is
+    empty.
+    """
+    if not content_staging_api:
+        raise RuntimeError("The required content_staging app is not installed")
+    user_clipboard = content_staging_api.get_user_clipboard_status(user.id)
+    if (
+        not user_clipboard or
+        user_clipboard.content.status != content_staging_api.StagedContentStatus.READY
+    ):
+        # Clipboard is empty or expired/error/loading
+        return None
+    block_type = user_clipboard.content.block_type
+    olx_str = content_staging_api.get_staged_content_olx(user_clipboard.content.id)
+    node = etree.fromstring(olx_str)
+    store = modulestore()
+    with store.bulk_operations(parent_key.course_key):
+        parent_xblock = store.get_item(parent_key)
+        runtime = parent_xblock.runtime
+        # Generate the new ID:
+        id_generator = ImportIdGenerator(parent_key.context_key)
+        def_id = id_generator.create_definition(block_type, user_clipboard.source_usage_key.block_id)
+        usage_id = id_generator.create_usage(def_id)
+        keys = ScopeIds(None, block_type, def_id, usage_id)
+        # parse_xml is a really messy API. We pass both 'keys' and 'id_generator' and, depending on the XBlock, either
+        # one may be used to determine the new XBlock's usage key, and the other will be ignored. e.g. video ignores
+        # 'keys' and uses 'id_generator', but the default XBlock parse_xml ignores 'id_generator' and uses 'keys'.
+        # For children of this block, obviously only id_generator is used.
+        xblock_class = runtime.load_block_type(block_type)
+        temp_xblock = xblock_class.parse_xml(node, runtime, keys, id_generator)
+        if xblock_class.has_children and temp_xblock.children:
+            raise NotImplementedError("We don't yet support pasting XBlocks with children")
+        temp_xblock.parent = parent_key
+        # Store a reference to where this block was copied from, in the 'copied_from_block' field (AuthoringMixin)
+        temp_xblock.copied_from_block = user_clipboard.source_usage_key
+        # Save the XBlock into modulestore. We need to save the block and its parent for this to work:
+        new_xblock = store.update_item(temp_xblock, user.id, allow_not_found=True)
```

Comment:
For Drag and Drop v2 XBlock, I'm getting the following error:
```python
2023-04-24 16:30:19,158 ERROR 4569 [django.request] [user None] [ip None] log.py:224 - Internal Server Error: /xblock/
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/views/decorators/http.py", line 40, in inner
    return func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/auth/decorators.py", line 21, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/util/json_request.py", line 55, in parse_json_into_request
    return view_function(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/block.py", line 258, in xblock_handler
    return _create_block(request)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/auth/decorators.py", line 21, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/util/json_request.py", line 55, in parse_json_into_request
    return view_function(request, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/block.py", line 723, in _create_block
    created_xblock = import_staged_content_from_user_clipboard(parent_key=usage_key, user=request.user)
  File "/edx/app/edxapp/edx-platform/cms/djangoapps/contentstore/views/helpers.py", line 349, in import_staged_content_from_user_clipboard
    new_xblock = store.update_item(temp_xblock, user.id, allow_not_found=True)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/mixed.py", line 84, in inner
    retval = func(field_decorator=strip_key_collection, *args, **kwargs)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/mixed.py", line 104, in wrapper
    return func(*args, **kwargs)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/mixed.py", line 795, in update_item
    return store.update_item(xblock, user_id, allow_not_found, **kwargs)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/split_draft.py", line 150, in update_item
    item = super().update_item(
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/split.py", line 1955, in update_item
    partitioned_fields = self.partition_xblock_fields_by_scope(descriptor)
  File "/edx/app/edxapp/edx-platform/xmodule/modulestore/split_mongo/split.py", line 3002, in partition_xblock_fields_by_scope
    result[field.scope][field.name] = field.read_from(xblock)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/xblock/fields.py", line 653, in read_from
    return self.__get__(xblock, xblock.__class__)  # pylint: disable=unnecessary-dunder-call
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/xblock/fields.py", line 506, in __get__
    field_data = xblock._field_data
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/xblock/mixins.py", line 230, in _field_data
    return self.runtime.service(self, 'field-data')
  File "/edx/app/edxapp/edx-platform/xmodule/x_module.py", line 1813, in service
    service = self._descriptor_system.service(block, service_name)
  File "/edx/app/edxapp/edx-platform/xmodule/x_module.py", line 1182, in service
    service = super().service(block=block, service_name=service_name)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/xblock/runtime.py", line 1125, in service
    raise NoSuchServiceError(f"Service {service_name!r} is not available.")
xblock.exceptions.NoSuchServiceError: Service 'field-data' is not available.
```

Reply:
This should be fixed now, via 8ce5ea8ba6aba7b58266fa23801ae4d3bd25e380